pkg/engine/planner/physical/optimizer.go

```diff
@@ -0,0 +1,183 @@
+package physical
+
+import (
+	"github.com/grafana/loki/v3/pkg/engine/internal/types"
+)
+
+// A rule is a tranformation that can be applied on a Node.
+type rule interface {
+	// apply tries to apply the transformation on the node.
+	// It returns a boolean indicating whether the transformation has been applied.
+	apply(Node) bool
+}
+
+// removeNoopFilter is a rule that removes Filter nodes without predicates.
+type removeNoopFilter struct {
+	plan *Plan
+}
+
+// apply implements rule.
+func (r *removeNoopFilter) apply(node Node) bool {
+	changed := false
+	switch node := node.(type) {
+	case *Filter:
+		if len(node.Predicates) == 0 {
+			r.plan.removeNode(node)
+			changed = true
+		}
+	}
+	return changed
+}
+
+var _ rule = (*removeNoopFilter)(nil)
+
+// predicatePushdown is a rule that moves down filter predicates to the scan nodes.
+type predicatePushdown struct {
+	plan *Plan
+}
+
+// apply implements rule.
+func (r *predicatePushdown) apply(node Node) bool {
+	changed := false
+	switch node := node.(type) {
+	case *Filter:
+		for i := 0; i < len(node.Predicates); i++ {
+			if ok := r.applyPredicatePushdown(node, node.Predicates[i]); ok {
+				changed = true
+				// remove predicates that have been pushed down
+				node.Predicates = append(node.Predicates[:i], node.Predicates[i+1:]...)
```

Comment:
[slices.Delete](https://pkg.go.dev/slices#Delete) is the more modern way of doing this; it does the same thing as this line but also handles [an edge case](https://cs.opensource.google/go/go/+/refs/tags/go1.24.1:src/slices/slices.go;l=229) to ensure obsolete elements can get garbage collected.